src/main/utility/interval_map.rs

```diff
@@ -286,43 +288,89 @@ mod tests {
         i.map(|x| x.end() - x.begin() + 1).sum()
     }
 
-    fn validate_map<V: Clone>(m: &IntervalMap<V>) {
+    fn validate_map<V: Clone>(m: &IntervalMap<V>) -> Result<(), String> {
         // Every interval is valid
         for i in m.keys() {
-            assert!(i.begin() <= i.end())
+            if i.begin() > i.end() {
+                return Err(format!("Invalid interval {:?}", i));
+            }
         }
 
         // Intervals don't overlap
         for (i1, i2) in m.keys().zip(m.keys().skip(1)) {
-            assert!(i1.end() < i2.begin());
+            if i1.end() >= i2.begin() {
+                return Err(format!("Overlapping intervals {:?} and {:?}", i1, i2));
+            }
         }
+
+        Ok(())
     }
 
-    // TODO: fuzz-test with this.
     fn insert_and_sanity_check(
         m: &mut IntervalMap<String>,
         begin: usize,
         end: usize,
         val: &str,
-    ) -> Vec<Mutation<String>> {
+    ) -> Result<Vec<Mutation<String>>, String> {
         let old_len_sum = interval_sum(m.keys());
-        let old_len = m.keys().count();
 
         // Do the insert.
         let mutations = m.insert(begin, end, val.to_string());
 
         // Validate general properties
-        validate_map(m);
+        validate_map(m)?;
 
         let new_len_sum = interval_sum(m.keys());
         let new_len = m.keys().count();
-        assert!(new_len_sum >= old_len_sum);
-        assert!(new_len_sum >= (end - begin + 1));
+        if new_len_sum < old_len_sum {
+            return Err(format!(
+                "length-sum shrunk from {} to {}",
+                old_len_sum, new_len_sum
+            ));
+        }
+        if !(new_len_sum >= (end - begin + 1)) {
+            return Err(format!(
+                "length-sum {} is smaller than inserted interval length {}",
+                new_len_sum,
+                end - begin + 1
+            ));
+        }
+        if new_len == 0 {
+            return Err(format!("new length is zero"));
+        }
 
-        assert!(new_len >= old_len);
```

Comment:
Was this `assert!(new_len >= old_len)` check intentionally removed? (just making sure it wasn't accidentally missed if it's still wanted)

Reply:
Yup! The fuzz testing tripped it, and I realized this was an incorrect assertion. the new number of intervals can be fewer than the old one, when the inserted interval removes multiple old intervals.

Reply:
Ah okay that makes sense!